File: rotting-oranges.js
// https://leetcode.com/problems/rotting-oranges/
// https://leetcode.com/submissions/detail/331917062/
/**
 * @param {number[][]} grid
 * @return {number}
 **/
var orangesRotting = function(grid) {
    const rows = grid.length;
    const cols = grid[0].length;
    const dirs = [[1, 0], [0, 1], [0, -1], [-1, 0]];
    const queue = [];

    var minutes = 0;
    let population = rows * cols;
    let freshOrangeCount = 0;

    // Find all rotten oranges....
    grid.forEach((row, r) => {
        row.forEach((orange, c) => {
          if (orange === 2) {
            queue.push([r, c]);
          }else if (orange ===  1){
              freshOrangeCount++
          }
        });
    });

    // no fresh oranges at minute 0
    if(freshOrangeCount==0) return 0;

    while (queue.length) {

        // console.log(`minutes = ${minutes} : queue = ${queue.toString()}`)

        for (let i = queue.length; i > 0; --i) {
            const rottenOrange = queue.shift();

            // Loop through all possible directions
            dirs.forEach(dir => {
                const target = [rottenOrange[0] + dir[0], rottenOrange[1] + dir[1]];

                    // If orange exists in grid and orange is fresh
                    // rot it
                    if (
                          target[0] >= 0 &&
                          target[0] < rows &&
                          target[1] >= 0 &&
                          target[1] < cols &&
                          grid[target[0]][target[1]] === 1
                    ) {
                      queue.push(target);
                      // Rot Orange
                      grid[target[0]][target[1]] = 2;
                    }
                });
            }
        minutes++;
        }

    let freshOrangeExists = false;
    // check if any fresh orange is there
    for(let i=0; i<rows; i++){
        for(let j=0; j<cols; j++){
            if (grid[i][j] === 1) {
                console.log(`Fresh Orange Found at last`)
                freshOrangeExists = true;
                break;
            }
        }
    }

    return freshOrangeExists ? -1 : minutes-1;
}

File: python/Matrix/rotting-oranges.py
# https://leetcode.com/problems/rotting-oranges/

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        good_oranges = 0
        queue = []
        minute = 0
        dirs = [[1,0], [-1,0], [0,1], [0,-1]]
        
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    good_oranges+=1
                elif grid[i][j] == 2:
                    queue.append([i,j])
    
        if good_oranges == 0:
            return 0
        
        if not queue:
            return -1
        
        while queue:     
            for _ in range(len(queue)):
                [i,j] = queue.pop(0)

                for dir_ in dirs:
                    target_i = i + dir_[0]
                    target_j = j + dir_[1]

                    if 0 <= target_i < len(grid) and 0 <= target_j < len(grid[0]) and grid[target_i][target_j] == 1:
                        grid[target_i][target_j] = 2
                        queue.append([target_i, target_j])
            
            minute+=1
                
        fresh_orange_exist = False
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    fresh_orange_exist = True
                    break
        
        return -1 if fresh_orange_exist else minute-1